tests/bsim/bluetooth/host/misc/unregister_conn_cb/src/main.c

```diff
@@ -0,0 +1,211 @@
+}
+
+static struct bt_conn_cb conn_callbacks = {
+	.connected = connected,
+	.disconnected = disconnected,
+};
+
+void device_found(const bt_addr_le_t *addr, int8_t rssi, uint8_t type, struct net_buf_simple *ad)
+{
+	char addr_str[BT_ADDR_LE_STR_LEN];
+	int err;
+
+	if (g_conn != NULL) {
+		printk("g_conn != NULL\n");
+		return;
+	}
+
+	/* We're only interested in connectable events */
+	if (type != BT_HCI_ADV_IND && type != BT_HCI_ADV_DIRECT_IND) {
+		printk("type not connectable\n");
+		return;
+	}
+
+	bt_addr_le_to_str(addr, addr_str, sizeof(addr_str));
+	printk("Device found: %s (RSSI %d)\n", addr_str, rssi);
+
+	printk("Stopping scan\n");
+	err = bt_le_scan_stop();
+	if (err != 0) {
+		FAIL("Could not stop scan: %d");
+		return;
+	}
+
+	err = bt_conn_le_create(addr, BT_CONN_LE_CREATE_CONN, BT_LE_CONN_PARAM_DEFAULT, &g_conn);
+	if (err != 0) {
+		FAIL("Could not connect to peer: %d", err);
+	}
+	printk("%s: connected to found device\n", __func__);
+}
+
+static void test_peripheral_main(void)
+{
+	int err;
+	const struct bt_data ad[] = {
+		BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR))};
+
+	err = bt_enable(NULL);
+	if (err != 0) {
+		FAIL("Bluetooth init failed (err %d)\n", err);
+		return;
+	}
+
+	printk("Bluetooth initialized\n");
+
+	bt_conn_cb_register(&conn_callbacks);
+
+	err = bt_le_adv_start(BT_LE_ADV_CONN_NAME, ad, ARRAY_SIZE(ad), NULL, 0);
+	if (err != 0) {
+		FAIL("Advertising failed to start (err %d)\n", err);
+		return;
+	}
+
+	printk("Advertising successfully started\n");
+
+	WAIT_FOR_FLAG(flag_is_connected);
+
+	WAIT_FOR_FLAG_UNSET(flag_is_connected);
+
+	bt_conn_cb_unregister(&conn_callbacks);
+
+	k_sleep(K_SECONDS(1));
+	err = bt_disable();
+	if (err != 0) {
+		FAIL("Bluetooth disable failed (err %d)\n", err);
+		return;
+	}
+
+	printk("Bluetooth successfully disabled\n");
+
+	PASS("Peripheral device passed\n");
+}
+
+static void test_central_main(void)
+{
+	int err;
+
+	err = bt_enable(NULL);
+	if (err != 0) {
+		FAIL("Bluetooth discover failed (err %d)\n", err);
+	}
+	printk("Bluetooth initialized\n");
+	bt_conn_cb_register(&conn_callbacks);
+	/* Connect to peer device after conn_callbacks registered*/
+	err = bt_le_scan_start(BT_LE_SCAN_PASSIVE, device_found);
+	if (err != 0) {
+		FAIL("Scanning failed to start (err %d)\n", err);
+	}
+
+	printk("Scanning successfully started\n");
+
+	WAIT_FOR_FLAG(flag_is_connected);
+
+	err = bt_conn_disconnect(g_conn, 0x13);
+
+	if (err != 0) {
+		FAIL("Disconnect failed (err %d)\n", err);
+		return;
+	}
+
+	WAIT_FOR_FLAG_UNSET(flag_is_connected);
+	bt_conn_cb_unregister(&conn_callbacks);
+	/* Reconnect to the device after conn_callbacks unregistered */
+	err = bt_le_scan_start(BT_LE_SCAN_PASSIVE, device_found);
+	if (err != 0) {
+		FAIL("Scanning failed to start (err %d)\n", err);
+	}
+	printk("Scanning successfully started\n");
+
+	k_sleep(K_SECONDS(1));
+	/* flag_is_connected not set means no conn_callbacks being called */
+	if (flag_is_connected) {
```

Comment:
you should add a conn-foreach to find out if you are really connected here. In addition to the flag not being set

Reply:
thanks Jori, have add below print in `device_found`() to indicate that connection did happen:
`printk("%s: connected to found device\n", __func__);  Please check the pass log:
`
![image](https://github.com/zephyrproject-rtos/zephyr/assets/149985409/17e48b76-0d4a-4679-b415-ce2f87664c36)

is it ok in this case?

Reply:
that's good for debugging, but the script needs to actually fail (i.e. return something else than 0) if that doesn't happen.
You need a `FAIL()` call in the case you're not connected. Else the CI is never going to fail, someone will have to inspect the logs.

Reply:
good point, have added the fail logic, pls help to review again, thanks a lot!

Reply:
looks good now :+1: